runtime/src/main/java/org/corfudb/runtime/clients/ClientResponseHandler.java

```diff
@@ -0,0 +1,235 @@
+     * @param ctx The channel handler context.
+     * @return True if the message was handled successfully.
+     */
+    public boolean handle(ResponseMsg response, ChannelHandlerContext ctx) {
+        IClientRouter router = client.getRouter();
+        long requestId = response.getHeader().getRequestId();
+        PayloadCase payloadCase = response.getPayload().getPayloadCase();
+
+        if (payloadCase.equals(PayloadCase.SERVER_ERROR)) {
+            ErrorCase errorCase = response.getPayload().getServerError().getErrorCase();
+            if (errorHandlerMap.containsKey(errorCase)) {
+                try{
+                    // TODO: should we remove this clause?
+                    Object ret = errorHandlerMap.get(errorCase).handle(response, ctx, router);
+                    if (ret != null) {
+                        router.completeRequest(requestId, ret);
+                    }
+                } catch (Throwable e) {
+                    router.completeExceptionally(requestId, e);
+                }
+            }
+        } else if (handlerMap.containsKey(payloadCase)) {
+            try {
+                Object ret = handlerMap.get(payloadCase).handle(response, ctx, router);
+                if (ret != null) {
+                    router.completeRequest(requestId, ret);
+                }
+            } catch (Throwable e) {
+                // TODO: same reason, should we remove this clause?
+                router.completeExceptionally(requestId, e);
+            }
+            return true;
+        }
+        return false;
+    }
+
+    public ClientResponseHandler generateHandlers(@NonNull final MethodHandles.Lookup caller,
+                                                  @NonNull final Object o) {
+        Arrays.stream(o.getClass().getDeclaredMethods())
+                .filter(method -> method.isAnnotationPresent(ResponseHandler.class))
+                .forEach(method -> {
+                    ResponseHandler handler = method.getAnnotation(ResponseHandler.class);
+                    if (!method.getParameterTypes()[0]
+                            .isAssignableFrom(ResponseMsg.class)) {
+                        throw new RuntimeException("Incorrect message type, expected "
+                                + ResponseMsg.class + " but provided "
+                                + method.getParameterTypes()[0]);
+                    }
+                    if (handlerMap.containsKey(handler.type())) {
+                        throw new RuntimeException("Handler for " + handler.type()
+                                + " already registered!");
+                    }
+                    // convert the method into a Java8 Lambda for maximum execution speed...
+                    try {
+                        if (Modifier.isStatic(method.getModifiers())) {
+                            MethodHandle mh = caller.unreflect(method);
+                            MethodType mt = mh.type().changeParameterType(0, ResponseMsg.class);
```

Comment:
You don't need to change the parameter type here. This what needed in the old implementation to change the instances of `CorfuPayloadMsg` (which extends `CorfuMsg`) to `CorfuMsg`. Here, everything will just be a `ResponseMsg` with nothing extending it.